src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.RSA.cs

```diff
@@ -0,0 +1,168 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+using System.Security.Cryptography.Apple;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static partial class AppleCrypto
+    {
+        [DllImport(Libraries.AppleCryptoNative, EntryPoint = "AppleCryptoNative_RsaGenerateKey")]
+        internal static extern int RsaGenerateKey(
+            int keySizeInBits,
+            out SafeSecKeyRefHandle pPublicKey,
+            out SafeSecKeyRefHandle pPrivateKey,
+            out int pOSStatus);
+
+        [DllImport(Libraries.AppleCryptoNative, EntryPoint = "AppleCryptoNative_RsaEncryptOaep")]
+        private static extern int RsaEncryptOaep(
+            SafeSecKeyRefHandle publicKey,
+            byte[] pbData,
+            int cbData,
+            PAL_HashAlgorithm mgfAlgorithm,
+            out SafeCFDataHandle pEncryptedOut,
+            out SafeCFErrorHandle pErrorOut);
+
+        [DllImport(Libraries.AppleCryptoNative, EntryPoint = "AppleCryptoNative_RsaEncryptPkcs")]
+        private static extern int RsaEncryptPkcs(
+            SafeSecKeyRefHandle publicKey,
+            byte[] pbData,
+            int cbData,
+            out SafeCFDataHandle pEncryptedOut,
+            out SafeCFErrorHandle pErrorOut);
+
+        [DllImport(Libraries.AppleCryptoNative, EntryPoint = "AppleCryptoNative_RsaDecryptOaep")]
+        private static extern int RsaDecryptOaep(
+            SafeSecKeyRefHandle publicKey,
+            byte[] pbData,
+            int cbData,
+            PAL_HashAlgorithm mgfAlgorithm,
+            out SafeCFDataHandle pEncryptedOut,
+            out SafeCFErrorHandle pErrorOut);
+
+        [DllImport(Libraries.AppleCryptoNative, EntryPoint = "AppleCryptoNative_RsaDecryptPkcs")]
+        private static extern int RsaDecryptPkcs(
+            SafeSecKeyRefHandle publicKey,
+            byte[] pbData,
+            int cbData,
+            out SafeCFDataHandle pEncryptedOut,
+            out SafeCFErrorHandle pErrorOut);
+
+        internal static byte[] RsaEncrypt(
+            SafeSecKeyRefHandle publicKey,
+            byte[] data,
+            RSAEncryptionPadding padding)
+        {
+            int ret;
+            SafeCFErrorHandle error;
+            SafeCFDataHandle encrypted;
+
+            if (padding == RSAEncryptionPadding.Pkcs1)
+            {
+                ret = RsaEncryptPkcs(publicKey, data, data.Length, out encrypted, out error);
+            }
+            else
+            {
+                ret = RsaEncryptOaep(
+                    publicKey,
+                    data,
+                    data.Length,
+                    PalAlgorithmFromAlgorithmName(padding.OaepHashAlgorithm),
+                    out encrypted,
+                    out error);
+            }
+
+            using (error)
```

Comment:
They can be SafeHandle objects wrapping IntPtr.Zero, but they themselves can't be null (to C#) without a terrible regression in the marshalling layer.